src/MDOP.jl

```diff
@@ -87,3 +220,134 @@ function beta_statistic(Y::Dataset, s::Vector; τs = 0:50 , w::Int = 1)
     β = mean(log10.(Δϕ); dims=1)
     return vec(β)
 end
+
+
+"""
+    embed2(Y::Dataset,s::Vector,τ::Int) -> Y_new
+adds the `τ` lagged values of the time series `s` as additional component to the
+embedding vector `Y`.
+"""
+function embed2(Y::Dataset,s::Vector{Float64},τ::Int)
+    N = size(Y,1)   # length of input trajectory
+    NN = size(Y,2)  # dimensionality of input trajectory
+    M = N - τ       # length of output trajectory
+
+    # preallocation
+    Y_new = zeros(M,NN+1)
+    # fill vector up until index M
+    @inbounds for i = 1:NN
+        Y_new[:,i] = Y[1:M,i]
+    end
+    # add lagged component of s
+    Y_new[:,NN+1] = s[1+τ:N]
+
+    return Dataset(Y_new)
+
+end
+
+"""
+    embed2(y::Vector,s::Vector,τ::Int) -> Y_new
+adds the `τ` lagged values of the time series `s` as additional component to the
+time series `y`.
+"""
+function embed2(y::Vector{Float64},s::Vector{Float64},τ::Int)
+    N = size(y,1)   # length of input trajectory
+    M = N - τ       # length of output trajectory
+
+    # preallocation
+    Y_new = zeros(M,2)
+    # fill vector up until index M
+    Y_new[:,1] = y[1:M]
+    # add lagged component of s
+    Y_new[:,2] = s[1+τ:N]
+
+    return Dataset(Y_new)
+
+end
```

Comment:
There is a simple way to make this function and it is not only more efficient, but also much more concise:
```julia
function embed2(y, s, τ)
    N = length(y)
    @assert N == length(s)
    M = N - τ
    return Dataset(view(y, 1:M), view(s, τ+1:N))
end
```

Some helpful tips: in the line `Y_new[:,1] = y[1:M]`, the command `y[1:M]` creates a **copy** of `y`. You should instead use the `view` function. 
Then, you can directly create a `Dataset` from input vectors: `Dataset(x, y, z, s, ...)` there is no reason to first make a matrix.

Reply:
In addition, I find the name `embed2` unintuitive: it doesn't seem to indicate what the function does, and it is also too similar with the existing `embed` and thus can be confused... Can we try to come up with a better name?

Reply:
(also notice that the comments in the original function are not helpful because it is clear from the source code what is happening. Comments should *add* something to make the source code clearer. Otherwise, they only add an additional line of code and make the source file larger ;) )

Reply:
The `view`-function is very handy, indeed!

Reply:
It seems, however, it is not possible to "slice" bigger pieces from a `Dataset`. I'll get error-messages on this one, when `y` is a Dataset:
```julia
function embed2(y, s, τ)
    N = length(y)
    @assert N == length(s)
    M = N - τ
    return Dataset(view(y, 1:M), view(s, τ+1:N))
end
```
> MethodError: no method matching view(::Dataset{1,Float64}, ::UnitRange{Int64}, ::Colon)
Closest candidates are:
  view(!Matched::StructArrays.StructArray{T,N,C,I} where I, ::Any...) where {T, N, C} at /Users/hkraemer/.julia/packages/StructArrays/OtfvU/src/structarray.jl:154
  view(!Matched::AbstractArray, ::Any...) where N at subarray.jl:155
  view(!Matched::AbstractUnitRange, ::AbstractUnitRange{var"#s91"} where var"#s91"<:Integer) at subarray.jl:167
  ...
embed2(::Dataset{1,Float64}, ::Array{Float64,1}, ::Int64) at MDOP.jl:235
MDOP(::Array{Float64,1}; τs::UnitRange{Int64}, w::Int64, fnn_thres::Float64, r::Int64, β_condition::Bool) at MDOP.jl:102
(::DelayEmbeddings.var"#MDOP##kw")(::NamedTuple{(:τs, :w, :β_condition),Tuple{UnitRange{Int64},Int64,Bool}}, ::typeof(MDOP), ::Array{Float64,1}) at MDOP.jl:56
top-level scope at test_MDOP.jl:142
include_string(::Function, ::Module, ::String, ::String) at loading.jl:1088

I could do:
```julia
function embed2(Y::Dataset{D,T}, s::Vector{T}, τ::Int) where {D, T<:Real}
    N = length(Y)
    @assert N == length(s)
    M = N - τ
    return Dataset(view(Y[:], 1:M), view(s, τ+1:N))
end
```
, but then `Dataset`-function does not know what to do. Finally,

```julia
function embed2(Y::Dataset{D,T}, s::Vector{T}, τ::Int) where {D, T<:Real}
    N = length(Y)
    @assert N <= length(s)
    M = N - τ
    return Dataset(hcat(view(Matrix(Y), 1:M, :), view(s, τ+1:N)))
end
```
works.

Reply:
Can you tell me **what** you want to do if `y` is a dataset? You want to get the values of which column at indices `1:M`? The last column? The first?

Reply:
And is the output of `embed2` always a dataset with 2 columns?  What does `embed2` do in general? If you describe what you want the function to do, perhaps we can find a simpler answer.

Reply:
Doing `Matrix(Y)` is bad: it copies the entire dataset into a matrix, and then copies that back into a dataset.

Reply:
As described in the `docstring`
```julia
"""
    embed_one_cycle(Y, s::Vector, τ::Int) -> Y_new
Add the `τ` lagged values of the time series `s` as additional component to the
time series `Y` (`Vector` or `Dataset`), in order to form a higher embedded
vector `Y_new`. The dimensionality of `Y_new` , thus, equals the dimensionality
of `Y+1`.
"""
```
this function just adds another `τ` lagged version of the input `s`, as an additional component to `Y`.

> Doing Matrix(Y) is bad: it copies the entire dataset into a matrix, and then copies that back into a dataset.

Hm. I tried a lot, but `view` does not work that well on `Dataset`'s I have the feeling...

Reply:
Yes, `view` doesn't work with `Dataset` in the way you would expect, because it is a vector of static vectors internally. It just "behaves" like a matrix.

Gimme some time and I'll write a performant version of `embed_one_cycle` and then tag you in the commit so you can see what I did.

Reply:
:+1: